stream.go

```diff
@@ -72,6 +79,19 @@ func (s *stream[T]) WaitNext() T {
 	return s.state.value
 }
 
+func (s *stream[T]) WaitNextCtx(ctx context.Context) (T, error) {
+	select {
+	// wait for changes
+	case <-s.Changes():
+		// advance to next value and return it
+		return s.Next(), nil
+
+	case <-ctx.Done():
+		var zeroVal T
+		return zeroVal, ctx.Err()
+	}
```

Comment:
The problem of the `select` statement is that, if multiple cases are true, there is no guarantee which case will be executed.

If you keep calling `WaitNextCtxt` with a cancelled context, there are times where this function will return changes if there are changes, and that is non-deterministic. I think we should always return an error if the context is cancelled.

My recommendation is to do the following, which will guarantee that changes won't be followed if the context is already closed.

```go
case <-s.Changes():
    select {
        case <-ctx.Done():
            var zeroVal T
            return zeroVal, ctx.Err()
        default:
            return s.Next(), nil
    }
case <-ctx.Done():
...
```

We would also need a test for this.

Reply:
That's a good point. What about "simply" changing it to the following?

```go
case <-s.Changes():
    return s.Next(), ctx.Err()
case <-ctx.Done():
...
```

The [description](https://pkg.go.dev/context#Context) for context.Err() says:
```
If Done is not yet closed, Err returns nil.
If Done is closed, Err returns a non-nil error explaining why:
Canceled if the context was canceled
or DeadlineExceeded if the context's deadline passed.
After Err returns a non-nil error, successive calls to Err return the same error.
```